### PR TITLE
Refactor pattern parsing with regex

### LINKED ISSUE #24: Refactor build_regex_from_pattern to use regex-based parsing for placeholders
The current implementation of the `build_regex_from_pattern` function in `crates/rstest-bdd/src/lib.rs` uses manual character-by-character parsing, including helper functions like `handle_opening_brace` and `handle_closing_brace`, to process placeholders and literals in the pattern string. This approach increases code complexity and can be error-prone, especially when handling escaped braces (e.g., `{{` and `}}`).

**Suggested Improvement:**
Refactor `build_regex_from_pattern` to use a single-pass, regex-based approach for splitting literals and placeholders. This would:
- Eliminate the need for manual brace-depth tracking and helper functions.
- Correctly handle escaped braces (`{{` and `}}`).
- Make the code shorter, clearer, and easier to maintain.

**Example Implementation:**
```rust
fn build_regex_from_pattern(pat: &str) -> String {
    // matches '{{' or '}}' or '{name[:type]}'
    let ph_re = Regex::new(r"\{\{|}}|{([^}:]+)(?::([^}]+))?}").unwrap();
    let mut regex = String::from("^");
    let mut last = 0;

    for cap in ph_re.captures_iter(pat) {
        let m = cap.get(0).unwrap();
        // literal between last..m.start
        regex.push_str(&regex::escape(&pat[last..m.start()]));

        match m.as_str() {
            "{{" => regex.push_str(r"\{"),
            "}}" => regex.push_str(r"\}"),
            _ => {
                let ty = cap.get(2).map(|m| m.as_str());
                let sub = type_subpattern(ty);
                regex.push('(');
                regex.push_str(sub);
                regex.push(')');
            }
        }

        last = m.end();
    }

    // trailing literal
    regex.push_str(&regex::escape(&pat[last..]));
    regex.push('$');
    regex
}
```

**Action Items:**
- Refactor `build_regex_from_pattern` to use the regex-based approach as outlined above.
- Remove the now-unnecessary helper functions and manual parsing logic.
- Ensure all existing tests pass and add new tests for edge cases (especially escaped braces).

This change will improve maintainability and clarity of the codebase.

Issue: https://github.com/leynos/rstest-bdd/issues/24

---

I created this issue for @leynos from https://github.com/leynos/rstest-bdd/pull/19#discussion_r2250217321.

<details>
<summary>Tips and commands</summary>

#### Getting Help

- [Contact our support team](mailto:support@sourcery.ai) for questions or feedback.
- Visit our [documentation](https://docs.sourcery.ai) for detailed guides and information.
- Keep in touch with the Sourcery team by following us on [X/Twitter](https://x.com/SourceryAI), [LinkedIn](https://www.linkedin.com/company/sourcery-ai/) or [GitHub](https://github.com/sourcery-ai).

</details>

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- replace manual brace scanning with regex-based pattern parsing
- document double-brace escaping and simplify placeholder tests

## Testing
- `make fmt`
- `make lint`
- `make test`
- `make markdownlint`
- `make nixie`

closes #24

------
https://chatgpt.com/codex/tasks/task_e_68a4286c6f5483228c27dcb7e51101a7

## Summary by Sourcery

Refactor the step pattern parser to use a single precompiled regex for splitting literal text and placeholders, replacing manual brace scanning with modular helper functions.

New Features:
- Replace manual brace scanning with regex-based pattern parsing
- Support double-brace escaping (‘{{’/’}}’) instead of backslash escapes
- Expand floating-point hints to accept decimals, scientific notation, and special values (NaN, inf, Infinity)

Enhancements:
- Split pattern builder into smaller functions (process_literal_segment, process_match, etc.) around a LazyLock placeholder regex
- Formalize placeholder syntax in code comments (name, type hint, whitespace rules, anchored matching)
- Update user guide to reflect the new escaping and placeholder rules

Documentation:
- Revise user-guide to document double-brace escaping, disallowed nested braces in placeholders, and updated type hint details

Tests:
- Convert placeholder tests to rstest parameterized cases
- Introduce helper `compiled` function to streamline pattern compilation in tests
- Add coverage for various float formats, brace-escaping scenarios, unbalanced/nested braces

### CODE REVIEW

File: crates/rstest-bdd/tests/placeholder_parsing.rs [hunk 1]
```diff
@@ -59,11 +59,17 @@ fn type_hint_uses_specialised_fragment() {
...
+            extract_placeholders(&pat, StepText::from(text)).is_some(),
+            "{text} should match f64",
+        );
+    }
 }
 
 #[test]
 fn handles_escaped_braces() {
-    let pat = StepPattern::from(r"literal \{ brace {v} \}");
+    let pat = StepPattern::from("literal {{ brace {v} }}");
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/tests/placeholder_parsing.rs:72]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

- Consider moving the placeholder‐matching regex into a static or lazy‐initialized variable to avoid recompiling it on every call to build_regex_from_pattern.
- Instead of panicking on Regex::new failure, you could initialize the placeholder regex once at startup (e.g. with once_cell or lazy_static) or propagate the error for more robust handling.
- The refactor no longer handles nested braces inside placeholders—if that change is intended, please document it or add a test to cover the expected behavior.

File: crates/rstest-bdd/src/lib.rs [hunk 2]
```diff
@@ -233,118 +239,73 @@ impl<'a> StepContext<'a> {
...
+    let mut last = 0;
+    let mut depth = 0i32;
+    for cap in ph_re.captures_iter(pat) {
+        let m = cap.get(0).unwrap_or_else(|| panic!("capture missing"));
+        if let Some(lit) = pat.get(last..m.start()) {
+            for ch in lit.chars() {
+                match ch {
+                    '{' => depth += 1,
+                    '}' => depth -= 1,
+                    _ => {}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/src/lib.rs:266]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

File: crates/rstest-bdd/src/lib.rs [hunk 3]
```diff
@@ -233,118 +243,142 @@ impl<'a> StepContext<'a> {
...
+        Some("u8" | "u16" | "u32" | "u64" | "u128" | "usize") => r"\d+",
+        Some("i8" | "i16" | "i32" | "i64" | "i128" | "isize") => r"[+-]?\d+",
+        // Accept integers, decimal forms with optional leading/trailing digits,
+        // scientific notation, and special values recognised by `FromStr`.
+        Some("f32" | "f64") => {
+            r"(?i:(?:[+-]?(?:\d+\.\d*|\.\d+|\d+)(?:[eE][+-]?\d+)?|nan|inf|infinity))"
+        }
+        _ => r".+?",
     }
 }
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/src/lib.rs:393]: Has this now been resolved?

File: crates/rstest-bdd/src/lib.rs [hunk 4]
```diff
@@ -233,118 +243,152 @@ impl<'a> StepContext<'a> {
...
+        last = m.end();
+    }
+    if let Some(tail) = pat.get(last..) {
+        regex.push_str(&regex::escape(tail));
     }
-    regex_source.push('$');
-    regex_source
+    regex.push('$');
+    regex
 }
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/src/lib.rs]: Has this now been resolved?

File: crates/rstest-bdd/src/lib.rs [hunk 5]
```diff
@@ -233,118 +239,133 @@ impl<'a> StepContext<'a> {
...
+        last = m.end();
     }
-    regex_source.push('$');
-    regex_source
+    if let Some(tail) = pat.get(last..) {
+        regex.push_str(&regex::escape(tail));
+    }
+    regex.push('$');
+    regex
 }
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/src/lib.rs]: Has this now been resolved?

File: crates/rstest-bdd/src/lib.rs [hunk 6]
```diff
@@ -233,118 +243,169 @@ impl<'a> StepContext<'a> {
...
+                    depth.saturating_add(usize::try_from(delta).expect("delta is non-negative"));
             }
-        };
-        i += advance;
+        } else {
+            #[expect(clippy::expect_used, reason = "delta is negative")]
+            {
+                depth = depth.saturating_sub(usize::try_from(-delta).expect("delta is negative"));
+            }
+        }
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/src/lib.rs]: Has this now been resolved?

**Comment by leynos** [comment]
@sourcery-ai review

File: crates/rstest-bdd/src/lib.rs [hunk 7]
```diff
@@ -233,119 +243,92 @@ impl<'a> StepContext<'a> {
...
+        #[expect(
+            clippy::expect_used,
+            reason = "placeholder regex guarantees UTF-8 bounds"
+        )]
+        let literal = pat.get(last..m.start()).expect("placeholder bounds");
+        for ch in literal.chars() {
+            match ch {
+                '{' => depth = depth.saturating_add(1),
+                '}' => depth = depth.saturating_sub(1),
+                _ => {}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/src/lib.rs:289]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/src/lib.rs:289]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

File: crates/rstest-bdd/tests/placeholder_parsing.rs [hunk 8]
```diff
@@ -1,13 +1,19 @@
...
 
+use rstest::rstest;
 use rstest_bdd::{StepPattern, StepText, extract_placeholders};
 
+#[expect(clippy::expect_used, reason = "test helper should fail loudly")]
+fn compiled(pattern: &'static str) -> StepPattern {
+    let pat = StepPattern::from(pattern);
+    pat.compile().expect("failed to compile pattern");
+    pat
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd/tests/placeholder_parsing.rs:11]: Has this now been resolved?
